### PR TITLE
Add CLI flag for background console redirection

### PR DESCRIPTION
**Pull request by QuiCM**
Requested a few times:
https://github.com/Pryaxis/TShock/issues/2003
https://github.com/Pryaxis/TShock/issues/1450

Draft until I (or someone else) actually tests this

### CODE REVIEW

**Comment by marco-jsf** [comment]
Any chance that this will make it into the next release? Is there something I can do to help?

File: TerrariaServerAPI/TerrariaApi.Server/HookManager.cs [hunk 1]
```diff
@@ -55,6 +55,13 @@ public void AttachOTAPIHooks(string[] args)
 			Hooking.ServerHooks.AttachTo(this);
 			Hooking.WiringHooks.AttachTo(this);
 			Hooking.WorldHooks.AttachTo(this);
+
+			//Run after Hooking.ServerHooks.AttachTo(this) to override the Hooks.Command.StartCommandThread hook
+			if (Console.IsInputRedirected && args.Any(x => x == "-redirect"))
+			{
+				ServerApi.LogWriter.ServerWriteLine($"TerrariaServer is running in the background. Input has been re-enabled.", TraceLevel.Info);
```
Comment (hakusaro) [inline TerrariaServerAPI/TerrariaApi.Server/HookManager.cs]: This message is just a bit unclear, at least to me.

```suggestion
				ServerApi.LogWriter.ServerWriteLine($"TerrariaServer is running in the background, but input was re-enabled by command line flag.", TraceLevel.Info);
```

**Review by hakusaro — APPROVED** [review]
lgtm, though I don't really have a good means to test this right now for various reasons. I think this should just get merged and kicked out because we can just fix a bug with it after pre-release.

**Comment by QuiCM** [comment]
> 
> 
> lgtm, though I don't really have a good means to test this right now for various reasons. I think this should just get merged and kicked out because we can just fix a bug with it after pre-release.

Kicking it out the door doesn't change the situation with not having a feasible means to test.
I'd rather be stuck with a potential feature in test than a non-functional feature in live

**Comment by hakusaro** [comment]
> Kicking it out the door doesn't change the situation with not having a feasible means to test.

Pre-release? Either it currently works the way it does (broken, redirected input doesn’t work) or it works (we fixed it). This isn’t going to make the server more broken. I don’t know the exact use case people have so I don’t know whether or not I can judge it as fulfilling the goal it attempts to fulfill for the people that want it.

**Comment by Rarisma** [comment]
So I'm the use case here,
I'm designing a [server manager](https://github.com/Rarisma/Rarismas-Server-Manager)
Currently I can only support TShock and the support is still shakey since I had to write a plugin which basically checks a file, and runs the command inside the file.

I tried a build with this fork, -redirect worked, with the console aknowledging that its there however it seems like it's not working when I send a command StandardInput.WriteLine and .Flush it didn't run it. This is probably an issue on my end however if someone has tested this feature, can you attach an example of how to use this feature.

**Comment by hakusaro** [comment]
@Rarisma just fwiw, did you play with something like using REST to issue commands? @DeathCradle do you have any idea how this might work?

**Comment by Rarisma** [comment]
No, I used my own program and modified it to send commands directly instead of using a helper plugin, this was why I'm unsure if its actually the build that doesn't work because it might be my program that doesn't work because TShock understands the --redirect flag

**Comment by SignatureBeef** [comment]
> @Rarisma just fwiw, did you play with something like using REST to issue commands? @DeathCradle do you have any idea how this might work?

i do somewhat, i've done something like a server manager many years ago but kinda lost that code. if i can find some time over the weekend ill see if i can replicate something, as i think there used to be a gotcha with Console.Read/Line in older frameworks which might be why that code was needed, but maybe not needed anymore. kinda depends if we have a working test case to confirm i guess; ill have to use windows to test the above repo (i primarily use osx) or come up with my own so if you have some test code for osx to save time id appreciate it.

**Comment by QuiCM** [comment]
Resolved under #195

**Pull request closed without merging**